conclave-enclave/src/main/kotlin/com/r3/conclave/enclave/internal/StreamEnclaveHostInterface.kt

```diff
@@ -0,0 +1,189 @@
+package com.r3.conclave.enclave.internal
+
+import com.r3.conclave.common.internal.*
+import com.r3.conclave.mail.internal.writeInt
+import com.r3.conclave.utilities.internal.getAllBytes
+import java.io.InputStream
+import java.io.OutputStream
+import java.nio.ByteBuffer
+import java.util.concurrent.ArrayBlockingQueue
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.Executors
+
+class StreamEnclaveHostInterface(
+        private val outputStream: OutputStream,     // Messages going to the host
+        private val inputStream: InputStream,       // Messages arriving from the host
+        maximumConcurrentCalls: Int
+) : EnclaveHostInterface() {
+    private val callExecutor = Executors.newFixedThreadPool(maximumConcurrentCalls)
+
+    /** On startup, send the maximum number of concurrent calls to the host. */
+    init {
+        outputStream.writeInt(maximumConcurrentCalls)
+    }
+
+    private val receiveLoop = object : Runnable {
+        private var done = false
+
+        /** Receive messages in a loop and send them to the appropriate call context. */
+        override fun run() {
+            while (!done) {
+                when (StreamCallInterfaceSignal.fromByte(inputStream.read().toByte())) {
+                    StreamCallInterfaceSignal.MESSAGE -> handleMessageCommand()
+                    StreamCallInterfaceSignal.STOP -> handleStopCommand()
+                }
+            }
+        }
+
+        /** Send the received message to the appropriate call context. */
+        private fun handleMessageCommand() {
+            val message = StreamCallInterfaceMessage.readFromStream(inputStream)
+            val callContext = enclaveCallContexts.computeIfAbsent(message.hostThreadID) {
+                val newCallContext = EnclaveCallContext(message.hostThreadID)
+                callExecutor.execute(newCallContext)
+                newCallContext
+            }
+            callContext.enqueMessage(message)
+        }
+
+        /** The host has told us there will be no more messages, shut everything down. */
+        private fun handleStopCommand() {
+            callExecutor.shutdown()     // Finish processing any outstanding calls
+            sendStopSignalToHost()     // Let the host know there will be no more messages
```

Comment:
Is this an ack back to the host?

Reply:
Nope, it's the command that stops the receive loop on the host side. It might seem odd that it works this way, but I've learned from past mistakes that this kind of thing works best when the stop signal comes down the same channel that whatever async process you are trying to stop blocks on (rather than using `interrupt()` or any similar mechanism).

Reply:
Because the host receive thread is blocking on the input stream from the enclave, so it makes sense that the sop command come from there.